### PR TITLE
Get-services API to talk to Pebble /v1/services

### PR DESCRIPTION
**Pull request by benhoyt**
This adds a `get_services` API to the Pebble client, and `get_services` to `model.Container` (along with a `get_service` helper to fetch a single service's status). This corresponds to the Pebble change https://github.com/canonical/pebble/pull/14

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -1068,6 +1068,26 @@ def get_plan(self) -> 'pebble.Plan':
...
+        """
+        return self._pebble.get_services(service_names)
+
+    def get_service(self, service_name: str) -> 'pebble.ServiceInfo':
+        """Get status information for a single named service.
+
+        Raises model error if service_name is not found.
+        """
+        services = self.get_services(service_name)
+        if not services:
```
Comment (hpidcock) [inline ops/model.py:1085]: I'm guessing an array of length 0 is falsey?

File: test/test_pebble.py [hunk 2]
```diff
@@ -935,6 +970,87 @@ def test_get_plan(self):
...
+        self.assertEqual(len(services), 2)
+        self.assertEqual(services[0].name, 'svc1')
+        self.assertEqual(services[0].startup, pebble.ServiceStartup.DISABLED)
+        self.assertEqual(services[0].current, pebble.ServiceStatus.INACTIVE)
+        self.assertEqual(services[1].name, 'svc2')
+        self.assertEqual(services[1].startup, pebble.ServiceStartup.ENABLED)
+        self.assertEqual(services[1].current, pebble.ServiceStatus.ACTIVE)
+
+        self.assertEqual(self.client.requests, [
+            ('GET', '/v1/services', {'names': ''}, None),
```
Comment (hpidcock) [inline test/test_pebble.py:1001]: I would have expected the client to not send names if they were not specified.

File: ops/pebble.py [hunk 3]
```diff
@@ -685,3 +730,14 @@ def get_plan(self) -> Plan:
...
+
+    def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
+        """Get the service status for the configured services.
+
+        If names is specified, only fetch the service status for the services
+        named.
+        """
+        if names is None:
+            names = []
+        result = self._request('GET', '/v1/services', {'names': ','.join(names)})
```
Comment (hpidcock) [inline ops/pebble.py:742]: I would have expected names not to be sent if empty.

**Review by hpidcock — APPROVED** [review]
LGTM

File: ops/model.py (hunk 1 above)
Comment (benhoyt) [reply ops/model.py:1085]: Yes, that's right. And it's idiomatic in Python to use that. See [comment in PEP 8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations): "For sequences, (strings, lists, tuples), use the fact that empty sequences are false"

File: test/test_pebble.py (hunk 2 above)
Comment (benhoyt) [reply test/test_pebble.py]: Yeah, true, that's a bit better -- fixed.

File: ops/pebble.py (hunk 3 above)
Comment (benhoyt) [reply ops/pebble.py]: Good call - updated.

**Review by jameinel — APPROVED** [review]

File: ops/pebble.py [hunk 4]
```diff
@@ -481,6 +481,51 @@ def __repr__(self) -> str:
...
+        self.name = name
+        self.startup = startup
+        self.current = current
+
+    @classmethod
+    def from_dict(cls, d: Dict) -> 'ServiceInfo':
+        """Create new object from dict parsed from JSON."""
+        return cls(
+            name=d['name'],
+            startup=ServiceStartup(d['startup']),
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply ops/pebble.py]: At first I thought no, just let the `ValueError` bubble up, but that means we can't extend this enum with new values without breaking potentially breaking charms. So I've updated it to just return the string if it isn't an enum value we recognize. Not perfect but definitely handles Pebble changes better.

File: ops/pebble.py [hunk 5]
```diff
@@ -481,6 +481,51 @@ def __repr__(self) -> str:
...
+            current=ServiceStatus(d['current']),
+        )
+
+    def __repr__(self):
+        return ('ServiceInfo('
+                'name={self.name!r}, '
+                'startup={self.startup}, '
```
[the remark this answers is not in the document]
Comment (benhoyt) [reply ops/pebble.py:548]: I think it's okay -- what would we do for a `__str__`, make up an arbitrary human-readable format like `'my-service active and enabled on startup'`. IMO if you wanted a specific format (e.g., for CLI output) you'd use `str.format` or similar explicitly. I haven't added explicit `__str__` for any of these other classes, so I've left it off for now.

As far as testing `__repr__`, I initially had tests for these, and they got really unwieldy. And then we decided to change them (remove newlines for logging), so all the tests would have had to change. Is the `__repr__` part of the contract? My thinking was probably not. I'll merge this now, but feel free to push back and I can add in another PR.

**Comment by jameinel** [comment]
I don't think `__repr__` is explicitly part of the contract, but it is
often useful to have a test and make sure it actually contains the
information in a readable form that is clear what you are trying to say.
(should there be quotes, etc).

**Pull request merged**